<div style="max-width: 700px; margin: 20px auto 30px; padding: 20px; text-align: center; border-bottom: 2px solid #e5e7eb;">
  <a href="01_autocorrelation_correlation.ipynb" style="display: inline-block; padding: 10px 24px; background: #f9fafb; color: #38549c; text-decoration: none; border-radius: 8px; font-weight: 600; font-size: 14px; border: 1px solid #e5e7eb; margin-right: 12px;">
    ← Part 1 - Autocorrelation and Correlation
  </a>
  <a href="03_seasonality.ipynb" style="display: inline-block; padding: 10px 24px; background: linear-gradient(180deg, #38549c 0%, #14276c 100%); color: white; text-decoration: none; border-radius: 8px; font-weight: 600; font-size: 14px; box-shadow: 0 4px 12px rgba(56,84,156,0.3);">
    Part 3 - Seasonality →
  </a>
</div>

<div style="max-width: 880px; margin: 20px auto 22px; padding: 0px; border-radius: 18px; border: 1px solid #e5e7eb; background: linear-gradient(180deg, #ffffff 0%, #f9fafb 100%); box-shadow: 0 8px 26px rgba(0,0,0,0.06); overflow: hidden;">
  <div style="padding: 34px 32px 14px; text-align: center; line-height: 1.38;">
    <div style="font-size: 13px; letter-spacing: 0.14em; text-transform: uppercase; color: #6b7280; font-weight: bold; margin-bottom: 5px;">
      Session #1 • Part 2
    </div>
    <!-- <div style="font-size: 29px; font-weight: 800; color: #14276c; margin-bottom: 4px;">
      Linear Time Series Models
    </div> -->
    <div style="font-size: 29px; font-weight: 800; color: #14276c; margin-bottom: 4px;">
      Stationarity and Unit Root Tests
    </div>
    <div style="font-size: 16.5px; color: #374151; font-style: italic; margin-bottom: 0;">
      Advanced Training School: Methods for Time Series
    </div>
  </div>
  <div style="background: none; text-align: center; margin: 30px 0 10px;">
    <img src="https://www.cemfi.es/images/Logo-Azul.png" alt="CEMFI Logo" style="width: 158px; filter: drop-shadow(0 2px 12px rgba(56,84,156,0.05)); margin-bottom: 0;">
  </div>
  <div style="font-family: 'Times New Roman', Times, serif; color: #38549c; text-align: center; font-size: 1.22em; font-weight: bold; margin-bottom: 0px;">
    Jesus Villota Miranda © 2025
  </div>
  <div style="font-family: 'Times New Roman', Times, serif; color: #38549c; text-align: center; font-size: 1em; margin-top: 7px; margin-bottom: 20px;">
    <a href="mailto:jesus.villota@cemfi.edu.es" style="color: #38549c; text-decoration: none; margin-right:8px;">jesus.villota@cemfi.edu.es</a>
    <span style="color:#9fa7bd;">|</span>
    <a href="https://www.linkedin.com/in/jesusvillotamiranda/" target="_blank" style="color: #38549c; text-decoration: none; margin-left:7px;">LinkedIn</a>
  </div>
</div>

<div style="max-width: 700px; margin: 20px auto 30px; padding: 24px 32px; border-radius: 18px; border: 1px solid #e5e7eb; background: linear-gradient(180deg, #ffffff 0%, #f9fafb 100%); box-shadow: 0 8px 26px rgba(0,0,0,0.06);">
<!-- <div style="font-size: 14px; letter-spacing: 0.1em; text-transform: uppercase; color: #6b7280; font-weight: bold; margin-bottom: 12px; text-align: center;">
  Session #1 • Part 2
</div> -->
<div style="font-size: 22px; font-weight: 700; color: #14276c; margin-bottom: 16px; text-align: center;">
  Table of Contents
</div>
<div style="font-size: 15px; color: #374151; line-height: 1.8;">
<div style="margin-bottom: 12px;"><strong>1.</strong> <a href="#section-1" style="color: #38549c; text-decoration: none;">Introduction to Stationarity</a></div>
<div style="margin-left: 20px; margin-bottom: 8px;">1.1 <a href="#section-1-1" style="color: #38549c; text-decoration: none;">Spurious Regression Simulation</a></div>
<div style="margin-bottom: 12px;"><strong>2.</strong> <a href="#section-2" style="color: #38549c; text-decoration: none;">Unit Root Tests: Theory and Methods</a></div>
<div style="margin-bottom: 12px;"><strong>3.</strong> <a href="#section-3" style="color: #38549c; text-decoration: none;">Application to IBEX-35</a></div>
<div style="margin-left: 20px; margin-bottom: 8px;">3.1 <a href="#section-3-1" style="color: #38549c; text-decoration: none;">Time Series Plots</a></div>
<div style="margin-left: 20px; margin-bottom: 8px;">3.2 <a href="#section-3-2" style="color: #38549c; text-decoration: none;">Correlograms</a></div>
<div style="margin-left: 20px; margin-bottom: 8px;">3.3 <a href="#section-3-3" style="color: #38549c; text-decoration: none;">DF-GLS Tests</a></div>
<div style="margin-left: 20px; margin-bottom: 8px;">3.4 <a href="#section-3-4" style="color: #38549c; text-decoration: none;">Interpreting Test Output</a></div>
</div></div>

<link rel="stylesheet" href="notebook_styles.css">

<link rel="stylesheet" href="notebook_styles.css">

<a id="section-1"></a>
<h2 class="styled-header">1) Introduction to Stationarity</h2>

**Stationarity** is a fundamental concept in time series analysis. A time series is **stationary** if its statistical properties (mean, variance, autocovariance) do not change over time.

**Why Stationarity Matters**:
- Standard statistical inference requires stationarity
- Non-stationary series can lead to **spurious regressions**
- Many time series models (ARMA) assume stationarity
- Non-stationary series often need transformation (differencing) before modeling

**Types of Stationarity**:

1. **Strict Stationarity**: Joint distribution invariant to time shifts
2. **Weak (Covariance) Stationarity**: 
   - Constant mean: $\mathbb{E}[Y_t] = \mu$ for all $t$
   - Constant variance: $\mathbb{V}[Y_t] = \sigma^2$ for all $t$
   - Autocovariance depends only on lag: $\gamma(t, t-h) = \gamma(h)$

<link rel="stylesheet" href="notebook_styles.css">

<a id="section-1-1"></a>
### 1.1 Spurious Regression Simulation

This simulation demonstrates why stationarity matters. We construct two independent random walks and regress one on the other to reproduce a classic spurious result. Formally, we draw i.i.d. shocks \(\varepsilon_t,\nu_t \sim \mathcal{N}(0,1)\) and build:

- \(x_t = x_{t-1}+\varepsilon_t\) with \(x_0=0\)
- \(y_t = y_{t-1}+\nu_t\) with \(y_0=0\)

Each series is non-stationary because its mean and variance drift over time. The regression produces a seemingly precise slope, large \(R^2\), and significant t-stat even though the two variables are unrelated—illustrating how unit roots invalidate standard inference unless the series are first differenced or cointegrated.

In [ ]:
* ---------------------------------------------------------------------------
* Spurious-regression demo: two random walks with no causal link
* ---------------------------------------------------------------------------
clear  // Clear all data from memory
set obs 250  // Create new dataset with 250 observations
set seed 20251121  // Set random number seed for reproducibility
gen t = _n  // Generate time index from 1 to 250
gen eps = rnormal()  // Generate random normal shocks for first random walk
gen nu = rnormal()  // Generate random normal shocks for second random walk
gen x_rw = sum(eps)  // Create first random walk by cumulating eps shocks
gen y_rw = sum(nu)  // Create second random walk by cumulating nu shocks
regress y_rw x_rw  // Run OLS regression of y_rw on x_rw (spurious regression)
display "Estimated slope: " _b[x_rw]  // Display estimated coefficient for x_rw
display "t-stat for x_rw: " (_b[x_rw]/_se[x_rw])  // Display t-statistic for x_rw coefficient
display "R-squared: " e(r2)  // Display R-squared from regression
predict yhat  // Generate predicted values from regression
predict resid, resid  // Generate residuals from regression
* Scatter plot of y_rw vs x_rw with navy circles and linear fit line in black
twoway (scatter y_rw x_rw, msymbol(circle) mcolor(navy)) ///
    (lfit y_rw x_rw, lcolor(black)) ///
    title("Independent random walks: regression fit") ///
    subtitle("High R-squared despite independence") ///
    name(sr_scatter, replace)
* Plot residuals over time as red line
twoway (line resid t, lcolor(red)) ///
    title("Residual drift from spurious regression") ///
    ytitle("Residual") xtitle("Time") ///
    name(sr_resid, replace)

<link rel="stylesheet" href="notebook_styles.css">

The `sr_scatter` graph shows a seemingly tight linear fit even though `x_rw` and `y_rw` are generated independently. The `sr_resid` graph makes the deeper problem explicit: the residuals wander over time just like the original series, so the usual inference (t-stats, R²) is invalid. This demonstrates why verifying stationarity (or cointegration) is essential before trusting the outputs of regressions with trending data.

<link rel="stylesheet" href="notebook_styles.css">

Having seen the danger of spurious regression, we now learn formal methods to test for stationarity.

<link rel="stylesheet" href="notebook_styles.css">

<a id="section-2"></a>
<h2 class="styled-header">2) Unit Root Tests: Theory and Methods</h2>

Before implementing unit root tests, we need to understand the mathematical framework underlying them. We'll first introduce the **Augmented Dickey-Fuller (ADF)** test, then the more powerful **DF-GLS** test.

#### The Augmented Dickey-Fuller (ADF) Test

The ADF test checks whether a time series has a unit root by estimating the following regression:

$$\Delta Y_t = \alpha + \delta Y_{t-1} + \sum_{i=1}^{p} \beta_i \Delta Y_{t-i} + \varepsilon_t$$

where:
- $\Delta Y_t = Y_t - Y_{t-1}$ is the first difference
- $\alpha$ is a constant (drift term)
- $\delta$ measures the degree of mean reversion
- The augmentation terms $\sum_{i=1}^{p} \beta_i \Delta Y_{t-i}$ account for serial correlation
- $\varepsilon_t$ is white noise

**Hypothesis Test**:
- **$H_0: \delta = 0$** (unit root exists → series is non-stationary)
- **$H_1: \delta < 0$** (series is stationary)

**Why $\delta < 0$ indicates stationarity**: If $\delta < 0$, the series exhibits **mean reversion** — when $Y_{t-1}$ is above its long-run mean, $\Delta Y_t$ tends to be negative (pulling it back down), and vice versa.

**The Non-Standard Distribution Problem**: Under the null hypothesis, the test statistic $t_\delta = \hat{\delta}/\text{SE}(\hat{\delta})$ does **not** follow a standard t-distribution. Instead, it follows the **Dickey-Fuller distribution**, which is left-skewed with critical values that are more negative than the standard normal (e.g., -2.86 at 5% vs. -1.96 for standard normal). This is why we need special critical value tables.

#### The DF-GLS Test (Elliott-Rothenberg-Stock, 1996)

The **DF-GLS test** improves upon the ADF test by using **Generalized Least Squares (GLS) detrending** before performing the unit root test. This approach has substantially better power, especially when the series is close to having a unit root.

**The GLS Transformation**:

Instead of testing the original series $Y_t$, DF-GLS first removes deterministic components (constant and/or trend) using GLS. The transformed series is:

$$\tilde{Y}_t = \begin{cases}
Y_1 & \text{if } t = 1 \\
Y_t - \bar{c} Y_{t-1} & \text{if } t > 1
\end{cases}$$

where $\bar{c} = 1 - \frac{7}{T}$ for the constant-only case (the `notrend` option), or $\bar{c} = 1 - \frac{13.5}{T}$ when including a trend. This transformation gives more weight to early observations where the series behavior is most informative about stationarity.

**After GLS detrending**, the DF-GLS test runs the ADF regression on the transformed series:

$$\Delta \tilde{Y}_t = \delta \tilde{Y}_{t-1} + \sum_{i=1}^{p} \beta_i \Delta \tilde{Y}_{t-i} + \varepsilon_t$$

The test statistic is:

$$\text{DF-GLS} = \frac{\hat{\delta}}{\text{SE}(\hat{\delta})}$$

**Key Advantages**:
- **Higher power**: Better at detecting stationarity when $\delta$ is close to zero (near unit root)
- **Less size distortion**: Fewer false rejections when the null is true
- **Asymptotically equivalent** to ADF but with better finite-sample properties
- **Standard choice** in modern applied econometrics

**Connection to Stata Implementation**:
- `dfgls varname, notrend`: Uses $\bar{c} = 1 - \frac{7}{T}$ (constant-only detrending)
- `dfgls varname`: Uses $\bar{c} = 1 - \frac{13.5}{T}$ (constant + trend detrending)
- `maxlag(p)`: Specifies maximum augmentation lags in $\sum_{i=1}^{p} \beta_i \Delta \tilde{Y}_{t-i}$

<link rel="stylesheet" href="notebook_styles.css">

With the theoretical foundation in place, we now apply these tests to real financial data.


<link rel="stylesheet" href="notebook_styles.css">

<a id="section-3"></a>
<h2 class="styled-header">3) Application to IBEX-35 Index and Returns</h2>

We test for stationarity in IBEX-35 data by comparing:

1. **Index levels** (prices) - expected to be **non-stationary**
2. **Returns** (first differences of log prices) - expected to be **stationary**


<link rel="stylesheet" href="notebook_styles.css">

<a id="section-3-1"></a>
### 3.1 Time Series Plots: Levels vs Returns


<link rel="stylesheet" href="notebook_styles.css">

With the theoretical foundation in place, we now apply these tests to real financial data.

<a id="section-3"></a>
<h2 class="styled-header">3) Application to IBEX-35 Index and Returns</h2>

We test for stationarity in IBEX-35 data by comparing:

1. **Index levels** (prices) - expected to be **non-stationary**
2. **Returns** (first differences of log prices) - expected to be **stationary**

<link rel="stylesheet" href="notebook_styles.css">

<a id="section-3-1"></a>
### 3.1 Time Series Plots: Levels vs Returns

In [ ]:
* ----------------------------------------------------------------------------
* Application to IBEX-35 data
* ----------------------------------------------------------------------------

* Load IBEX-35 daily data from processed file
global processed_data "../../data/processed"  // Define path to processed data folder
use "$processed_data/ibex35_daily.dta", clear  // Load IBEX-35 daily dataset

* Time series plots
* Plot IBEX-35 closing prices over time
tsline ibex35_close, ///
    title("IBEX-35 Index Levels") ///
    ytitle("Index") xtitle("Date") ///
    name(ibex_levels, replace)

* Plot IBEX-35 returns over time
tsline ribex35, ///
    title("IBEX-35 Daily Returns") ///
    ytitle("Returns (%)") xtitle("Date") ///
    name(ibex_returns, replace)

**Visual Inspection**:
- **Levels**: Show a clear **trend** (non-stationary - mean changes over time)
- **Returns**: Fluctuate around **zero** (stationary - constant mean)

This visual inspection suggests that levels are non-stationary while returns are stationary.

<link rel="stylesheet" href="notebook_styles.css">

<a id="section-3-2"></a>
### 3.2 Correlograms: Levels vs Returns

In [ ]:
* Correlograms
* Plot autocorrelation function for IBEX-35 levels up to lag 40
ac ibex35_close, lags(40) ///
    title("ACF: IBEX-35 Levels") ///
    name(acf_ibex_levels, replace)

* Plot autocorrelation function for IBEX-35 returns up to lag 40
ac ribex35, lags(40) ///
    title("ACF: IBEX-35 Returns") ///
    name(acf_ibex_ret, replace)

**ACF Patterns**:
- **Levels**: ACF decays **very slowly** (characteristic of non-stationary series)
- **Returns**: ACF decays **quickly** and stays within confidence bands (characteristic of stationary series)

The slow decay in levels indicates **long memory** and suggests a unit root process.

<link rel="stylesheet" href="notebook_styles.css">

<a id="section-3-3"></a>
### 3.3 DF-GLS Unit Root Tests: Implementation

In [4]:
* DF-GLS unit root tests
* Create continuous series for dfgls (requires no gaps)
* Clear any previous preserve state that may have been left incomplete
capture restore  // Silently restore if any preserve state exists
preserve  // Save current dataset state to restore later
drop if missing(ibex35_close)  // Remove observations with missing IBEX-35 closing prices
gen t = _n  // Generate continuous time index from 1 to N
tsset t  // Declare dataset as time series with variable t as time index
display _newline "DF-GLS Test for IBEX-35 Levels:"  // Print header for test results
display "{hline 60}"
dfgls ibex35_close, maxlag(3) notrend  // Run DF-GLS unit root test (display full output)
display _newline "Interpretation:"
display "  H0: Series has a unit root (non-stationary)"
display "  If test statistic < critical value → reject H0 (stationary)"
display "  If test statistic > critical value → fail to reject H0 (non-stationary)"
restore  // Restore dataset to state before preserve

preserve  // Save current dataset state to restore later
drop if missing(ribex35)  // Remove observations with missing IBEX-35 returns
gen t = _n  // Generate continuous time index from 1 to N
tsset t  // Declare dataset as time series with variable t as time index
display _newline _newline "DF-GLS Test for IBEX-35 Returns:"  // Print header for test results
display "{hline 60}"
dfgls ribex35, maxlag(3) notrend  // Run DF-GLS unit root test (display full output)
display _newline "Interpretation:"
display "  Returns should show negative test statistics well below critical values"
display "  → Reject H0 → Returns are stationary"
restore  // Restore dataset to state before preserve





(0 observations deleted)



Time variable: t, 1 to 8197
        Delta: 1 unit


DF-GLS Test for IBEX-35 Levels:

------------------------------------------------------------



Time variable: t, 1 to 8197
        Delta: 1 unit


DF-GLS Test for IBEX-35 Levels:

------------------------------------------------------------


DF-GLS test for unit root                Number of obs = 8,193
Variable: ibex35_close
Lag selection: User specified            Maximum lag   =     3

                             -------- Critical value ---------
 [lags]       DF-GLS mu           1%           5%          10%
--------------------------------------------------------------
      3           0.440       -2.580       -1.950       -1.626


DF-GLS test for unit root                Number of obs = 8,193
Variable: ibex35_close
Lag selection: User specified            Maximum lag   =     3

                             -------- Critical value ---------
 [lags]       DF-GLS mu           1%           5%      

<link rel="stylesheet" href="notebook_styles.css">

<a id="section-3-4"></a>
### 3.4 Understanding DF-GLS Test Output

The DF-GLS test output follows a simple logic. Here's how to interpret it:

**Decision Rule**: In unit root tests, we reject when the test statistic is large and negative.

$$\text{Decision Rule: If } |\text{Test Statistic}| > |\text{Critical Value}| \text{ → Reject } H_0$$

| Outcome | Test Statistic | Decision | Interpretation |
|---------|----------------|----------|-----------------|
| **Very negative** | e.g., -29.924 | Reject $H_0$ | **Stationary** ✓ |
| **Less negative** | e.g., -2.5 | Reject or borderline | Depends on significance level |
| **Close to zero/positive** | e.g., 0.440 | Fail to reject $H_0$ | **Non-stationary** ✗ |

**Reading the Critical Values Table**:

```
                         -------- Critical value ---------
 [lags]       DF-GLS mu       1%           5%          10%
--------------------------------------------------------------
      3           0.440     -2.580       -1.950       -1.626
```

These three critical values correspond to different significance levels (1%, 5%, 10%). If your test statistic is **more negative** than the critical value, you reject the null hypothesis at that significance level.

**IBEX-35 Levels**: Test Statistic = **0.440**
- 0.440 > -2.580, -1.950, -1.626 (all critical values)
- **Interpretation**: At **all significance levels**, we **fail to reject** $H_0$. The series is **non-stationary**.

**IBEX-35 Returns**: Test Statistic = **-29.924**
- -29.924 < -2.580, -1.950, -1.626 (all critical values)
- **Interpretation**: At **all significance levels**, we **reject** $H_0$. The series is **stationary**.

**Lag Selection**: The output shows information criteria (Opt lag, Min SIC, Min MAIC) that help validate the choice of `maxlag(3)`. The `maxlag` parameter tells Stata to test specifications with 1, 2, and 3 lags to account for serial correlation in residuals.

**The `notrend` Option**: For financial data like stock prices, we use `notrend` because prices are driven by random shocks (stochastic trend), not a fixed linear deterministic trend.

**Key Insight**: The dramatic difference between levels and returns demonstrates the **power of differencing**:

$$\text{Price levels: } P_t \text{ (non-stationary, } I(1)) \quad \text{Returns: } r_t = \Delta \ln(P_t) \text{ (stationary, } I(0))$$

This transformation is crucial because standard regression assumes stationarity, and non-stationary regressors can cause spurious correlations.